Alpha release channel and build architecture support

yougotwill opened the pull request:
- Added support for requesting alpha releases
- Added platform os and architecture query params
- Updates to the documentation replacing https://github.com/session-foundation/session-file-server/pull/1

All feedback and change requests welcome

yougotwill committed:
feat: add project key to pyproject.toml

yougotwill committed:
fix: align and add missing code blocks to README

yougotwill committed:
feat: updated README to point to docker Readme

formatted docker README

yougotwill committed:
fix: update git org to session-foundation in docker readme

yougotwill committed:
fix: updated github org to session-foundation

yougotwill committed:
feat: added alpha releases support in db

yougotwill committed:
feat: added support platform os, arch and release channel in session_version route

return alpha release information when we are using that channel

yougotwill committed:
fix: update request arg from releases to release_channel

yougotwill committed:
fix: add missing placeholders in account_version_checks

add release channel to return version number

yougotwill committed:
feat: updated docker readme with persisting the server pubkey for testing

yougotwill committed:
feat: git ignore vscode folder

yougotwill committed:
fix: store alpha build number in database instead of messing with the version code

we not get the correct releases depending on the channel we use

yougotwill committed:
fix: updated docker readme to mention you need the priv key file to be in bytes

yougotwill committed:
fix: rename variable for clarity in cleanup.py

alpha_version to alpha_build_number

jagerman commented on schema.pgsql:
```diff
@@ -45,23 +46,26 @@ CREATE VIEW versions AS
-INSERT INTO projects (name) VALUES ('oxen-io/session-desktop');
-INSERT INTO projects (name) VALUES ('oxen-io/session-android');
-INSERT INTO projects (name) VALUES ('oxen-io/session-ios');
+INSERT INTO projects (name) VALUES ('yougotwill/session-desktop');
```
Dev changed left in

yougotwill replied:
Good catch! Fixed

jagerman commented on fileserver/routes.py:
```diff
@@ -345,24 +353,39 @@ def get_session_version():
-            WHERE proj_name = %s ORDER BY version_code DESC""",
-            (project,),
-        )
+        if release_channel == "alpha":
+            # Fetch the latest alpha release version
+            cur.execute(
+                """
+                SELECT id, version, name, notes, alpharelease from alpharelease_versions
+                WHERE proj_name = %s ORDER BY version_code DESC""",
+                (project,),
+            )
+        else:
+            # Fetch the latest release version
+            cur.execute(
+                """
+                SELECT id, version, name, notes, alpharelease from release_versions
+                WHERE proj_name = %s ORDER BY version_code DESC""",
+                (project,),
+            )
```
Perhaps just use an f-string on the view name here to avoid most of the duplication (unless I'm missing some other difference?):
```suggestion
        # Fetch the latest release (or alpha release) version
        cur.execute(
            f"""
            SELECT id, version, name, notes, alpharelease
            FROM {'alpha' if release_channel == "alpha" else ''}release_versions
            WHERE proj_name = %s ORDER BY version_code DESC""",
            (project,),
        )
```

yougotwill replied:
You haven't missed anything and are correct. I have fixed it. 👍

jagerman commented on fileserver/routes.py:
```diff
@@ -345,24 +353,39 @@ def get_session_version():
+        alpha_build_number = row[4]
+
+        if alpha_build_number is not None:
+            release_version += f"-alpha.{alpha_build_number}"
```
Perhaps in the `versions` view we could change the `version` column from so that this is reflected directly in the view, rather than having to be manipulated here in the code:
```
version_code / 1000000 || '.' || version_code % 1000000 / 1000 || '.' || version_code % 1000 AS version,
```
to:
```
version_code / 1000000 || '.' || version_code % 1000000 / 1000 || '.' || version_code % 1000 ||
    (CASE WHEN alpharelease IS NOT NULL THEN '-alpha.' || alpharelease ELSE '')
    AS version,
```

yougotwill replied:
Oh cool. I have updated it. 🙏

jagerman commented on fileserver/routes.py:
```diff
@@ -317,6 +317,14 @@ def get_session_version():
         return error_resp(http.NOT_FOUND)
     project = "session-foundation/session-" + platform
 
+    platform_os = request.args.get("os")
+    platform_arch = request.args.get("arch")
+    release_channel = request.args.get("release_channel") if request.args.get("release_channel") else 'latest'
+
+    # Backwards compatibility from before apple silicon builds
+    if platform_os is not None and platform_os == 'darwin' and platform_arch is None:
+        platform_arch = 'x86_64'
+
```
I think that the change to include these values in requests + db storage warrants some more discussion (perhaps in a separate PR) quite apart from being able to retrieve the alpha version.

The more fields we have, the more possibility that some combination of those fields could pinpoint a specific user's activity.  (For example: if we added arm64 linux builds at some point, and there is only one linux arm64 user on the alpha release channel then that information would effectively allow you to tease out when that particular user is online).

yougotwill replied:
That is a good point. I will revert them from this PR and open a separate PR with just those changes for further discussion

yougotwill replied:
Turns out we don't need it at all for apple silicon builds. So unless we want that info probably don't need that second PR?

yougotwill committed:
fix: remove dev change to project info

yougotwill committed:
refactor: simplify version fetching logic in get_session_version function

yougotwill committed:
fix: remove alpha build number logic routes code and use versions view in schema instead

yougotwill committed:
fix: remove platform_os and platform_arch from account_version_checks

yougotwill committed:
fix: correct if syntax in version view definition

yougotwill committed:
fix: remove extra placeholder from account_version_checks insert statement

jagerman commented on fileserver/cleanup.py:
```diff
@@ -68,24 +68,32 @@ def periodic(signum):
                                     notes = EXCLUDED.notes
-                                    WHERE releases.prerelease != EXCLUDED.prerelease
+                                    WHERE releases.alpharelease < EXCLUDED.alpharelease
+                                        AND (releases.prerelease != EXCLUDED.prerelease
```
This `alpha < EXCLUDED.alpha AND (...)` doesn't seem right to me for a few reasons, but most importantly because comparisons with null values are always false, the `NULL < NULL AND (...)` here will always fail to update changed info (e.g. edited description, or changed prerelease status) of non-alpha-release builds.

I'm thinking we might instead need to tweak the version_code here to incorporate the alpha status.  Maybe something like:

```Python
1_000_000_000 * MAJOR + 1_000_000 * MINOR + 1_000 * PATCH + (999 if alpha_build_number is None else alpha_build_number)
```
so that non-alpha builds sort higher than alpha builds, and alpha builds have a unique version value.

But looking at that, I really don't remember why we use a computed version code at all, versus just having `vmajor SMALLINT NOT NULL`, `vminor SMALLINT NOT NULL`, `vpatch SMALLINT NOT NULL` and `valpha SMALLINT` and unique index on `(project_code, vmajor, vminor, vpatch, valpha)` and so going that way is another alternative.

yougotwill replied:
That is a good point. I attempted to do what you suggested with a different formula for the version_code. Yours looks good but actually I would prefer having the SMALLINTS. Are you happy for me to do that approach?

yougotwill replied:
Fixed by using the SMALLINTS solution

jagerman commented on pyproject.toml:
```diff
@@ -5,3 +5,11 @@ skip-magic-trailing-comma = true
+description = "Official Python implementation of the Session file server "
+license = {text = "GNU GPLv3"}
+readme = "README.md"
+requires-python = "==3.8.*"
```
```suggestion
requires-python = ">= 3.8"
```
(I think the production file server is currently 3.11, but anything 3.8+ should work)

yougotwill replied:
Makes sense to me.

yougotwill replied:
Fixed

yougotwill committed:
fix: update Python version requirement to allow newer versions

yougotwill committed:
fix: add .venv to .gitignore to exclude virtual environment files

yougotwill committed:
refactor: LICENSE.txt to LICENSE

yougotwill committed:
refactor:  use separate major, minor, patch, and alpha fields instead of version_code

yougotwill committed:
fix: replace deprecated app.logger.warn() with app.logger.warning() for improved logging consistency

yougotwill committed:
fix: use the correct capture group for alpha build numbers

yougotwill committed:
fix: correct indexing for release and prerelease version details in get_session_version

yougotwill committed:
fix: remove unnecessary whitespace in get_session_version function

yougotwill committed:
fix: readd project entries for session-android and session-ios in the database schema

jagerman commented on fileserver/routes.py:
```diff
@@ -313,10 +313,12 @@ def get_session_version():
         return error_resp(http.NOT_FOUND)
     project = "session-foundation/session-" + platform
 
+    release_channel = request.args.get("release_channel") if request.args.get("release_channel") else 'latest'
```
Perhaps `stable` rather than `latest`?

jagerman replied:
My thinking is that we might want `latest` to mean "give me the latest from any channel", and that could be useful if you are following alpha builds, so that you get alpha builds when there is something newer, but when the stable release is > the alpha release (i.e. when there is no current alpha release), you get the stable release instead.

jagerman replied:
Actually, looking at how this runs, that probably requires a fair amount of extra work so I'm fine leaving it as is.

yougotwill replied:
So I used "latest" because that is term used by our electron compiler ([electron-builder](https://www.electron.build/)). But Audric has already asked for the UI to use "stable" so I am happy to make the change here too.

~~Will update~~

Fixed

jagerman commented on schema.pgsql:
```diff
@@ -40,28 +43,33 @@ CREATE INDEX ON release_assets(release);
-CREATE VIEW release_versions AS SELECT * FROM versions WHERE NOT prerelease;
+CREATE VIEW release_versions AS SELECT * FROM versions WHERE NOT prerelease AND valpha IS NULL;
 CREATE VIEW prerelease_versions AS SELECT * FROM versions WHERE prerelease;
+CREATE VIEW alpharelease_versions AS SELECT * FROM versions WHERE valpha IS NOT NULL;
```
I think these views are counterproductive now that there are more than two channels.  Perhaps instead we could have just bake the channel into the `versions` view like this:
```SQL
CREATE VIEW versions AS
    SELECT
        releases.id AS id,
        -- other fields, etc.,
        CASE
            WHEN valpha IS NOT NULL THEN 'alpha'
            WHEN prerelease THEN 'prerelease'
            ELSE 'stable'
        END as channel
    FROM releases JOIN project ...
```
and then when querying, rather than altering which view we query from, instead of:
```
            SELECT id, vmajor, vminor, vpatch, valpha, version, name, notes
            FROM {'alpha' if release_channel == 'alpha' else ''}release_versions
            WHERE xyz
            ...
```
we would do:
```
            SELECT id, vmajor, vminor, vpatch, valpha, version, name, notes
            FROM versions
            WHERE channel = 'stable' AND xyz
            ...
```

Then you could do more interesting things like "give me the latest from any channel" by omitting the channel condition entirely, or "give me releases or prereleases" with a `WHERE channel IN ('stable', 'prerelease')`.

yougotwill replied:
Fixed

yougotwill committed:
fix: update default release_channel from 'latest' to 'stable' in session version handling

yougotwill committed:
fix: removed additional versions views and use channel property when making queries about different release types

jagerman committed:
Small fixes

- use get-with-default instead of getting twice
- fix unique version index to treat nulls as equivalent
- increase max size of account_version_checks.channel for potential
  future longer values (smaller varchar limits just constrain the value
  but don't save any space).
